### PR TITLE
Added public function `Mix.Tasks.Xref.calls/0` (#6967)

### LINKED ISSUE #6967: Mix.Xref.calls/0 - list all function calls in an Elixir project
## High Level Description

`Mix.Tasks.Xref` should provide a public function `Mix.Tasks.Xref.calls/0` that returns data about the function or macro calls in all files in the current project. For each function call, `Mix.Tasks.Xref.calls/0` should return the function that is called (module, function and arity), at least the file where it is called, the module where it is called and line where it is called.

This information could be presented in several ways:

#### As a Flat List (fully normalized)

In this case, the final result would be a flat list of function calls. Each row should be represented as a map, so that adding new keys in future versions doesn't break compatibility. Returning a list of tuples will break compatibility if new keys are added, because pattern matching on tuples is positional. The keys in the map should be:

* `callee` - A tuple containing the module, function and arity of the call
* `line` - Line where the function is called (integer)
* `file` - File where the function is called (binary)
* `caller_module` - Module where the function is called (atom) - this can in principle be deduced from the file, line and information derived from the BEAM debug chunks, but I think it would be convenient to have this information directly accessible here

I leave out of this map information about which function calls our function, because that's pretty easy to deduce accurately from the BEAM debug chunks, but that information could be included too. In that case two other keys could be included:

* `caller_function`
* `caller_arity`

These keys could be `nil` if the function is not called inside another function. Again, these two keys are probably out of the scope of Mix.Xref and should be handled by "userland" code. I include them here mostly for completeness.

This is the fully denormalized representation. It's useful to insert into a database to be queried later, or even "queried" with normal `Enum` functions, like `Enum.filter`. The disadvantage is that it might be inconvenient for some applications.

#### As a Map (denormalized)

Normalizing the data requires choosing an axis over which to aggregate the data. The logical choice is probably the called function. The functions module, name and arity could work as a key for a map with the following structure:

```
%{{m, f, a} => [%{line: line, file: file, caller_module: caller_module}, ...]}
```

This data supports querying only on a single axis ("Where is this function called?"), and would have to be denormalized for other queries. That's why I support the fully denormalized representation. It's preppty simple to turn it into this normalized representation anyway.

## Motivation

The Elixir compiler knows a lot about the interrelations between different parts of the code. Sharing that "knowledge" allows developper tools to be smarter.  Examples of such tools are:

* Source viewers, like my own [Guaxinim](https://github.com/tmbb/guaxinim) (demo here: https://tmbb.github.io/guaxinim/guaxinim.ex.html), which might have hyperlinks from the function calls to their definitions
* Other kinds of visualization tools. For example, there could be graph visualizations at higher level of granularity than what Mix.Xref provides. Those tools cold be implemented by the end users, which would take away work from the core team.
* Refactoring tools, which can tell the developer instantaneously where a certain function is called and possibly rename it.
* Better support of the "jump to definition" functionality in IDEs. Maybe this could be used by the Elixir language server?

## Implementation

This call list can be derived from the `compile_dispatches` and `runtime_dispatches` fields.

Currently, for Elixir 1.5, Guaxinim extracts these data from the compile manifests, so `Mix.Tasks.Xref `can take the same approach. The only overhead this brings is the need to update the `Mix.Tasks.Xref.calls/0` function as the internal representation evolves.

### PR DESCRIPTION
**Pull request by jgmchan**
The function will return a flat list of maps containing information on all the called functions.

Closes #6967

- [x] Write doco
- [x] Add Spec
~~- [ ] Work out how to support Umbrella apps (?)~~

### CODE REVIEW

**Comment by jgmchan** [comment]
I've updated the code according to the comments, squashed and rebased the branch to master.

File: lib/mix/lib/mix/tasks/xref.ex [hunk 1]
```diff
@@ -192,6 +192,61 @@ defmodule Mix.Tasks.Xref do
...
+    * `:line` - Line where the function is called (integer)
+    * `:file` - File where the function is called (binary)
+    * `:caller_module` - Module where the function is called (atom)
+
+  This function returns an empty list when used at the root of an umbrella
+  project because there is no compile manifest to extract the function call
+  information from. To get the function calls of each child in an umbrella,
+  execute the function at the root of each individual application.
+  """
+  @spec calls() :: [map]
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/xref.ex]: Can you please update the spec to be something like:

  @spec calls() :: [%{callee: {module, atom, arity}, caller_module: module, line: integer(), file: String.t]

Please don't forget to run the formatter after you do. :) Thanks!

Comment (jgmchan) [reply lib/mix/lib/mix/tasks/xref.ex]: Ok no problem!

**Comment by jgmchan** [comment]
Updated spec to be more specific. I think everything is ready now.

File: lib/mix/lib/mix/tasks/xref.ex [hunk 2]
```diff
@@ -197,10 +197,10 @@ defmodule Mix.Tasks.Xref do
 
   Each item in the list is a map with the following keys:
 
-    * `:callee` - A tuple containing the module, function and arity of the call
-    * `:line` - Line where the function is called (integer)
-    * `:file` - File where the function is called (binary)
-    * `:caller_module` - Module where the function is called (atom)
+    * `:callee` - a tuple containing the module, function, and arity of the call
+    * `:line` - an integer represeting the line where the function is called
```
Comment (jgmchan) [inline lib/mix/lib/mix/tasks/xref.ex]: Slight typo - `representing`

Comment (whatyouhide) [reply lib/mix/lib/mix/tasks/xref.ex]: Great catch, believe it or not my "n" key on my laptop has been acting up the last few days so this is far from the first time it happens 😄

Comment (jgmchan) [reply lib/mix/lib/mix/tasks/xref.ex]: 😆 That is unfortunate!

**Comment by whatyouhide** [comment]
Thanks @jgmchan! 💟

**Pull request merged**